server, tracing: Make traces redactable and default to "off"

dhartunian opened the pull request:
Multi-tenant deployments require tracing support
in order to help customers and engineers debug query performance
issues. This requires that trace information recorded on the KV layer
be redacted prior to being transmitted to the tenant in order to
guard against data leakage through the multi-tenant abstractions.

This change makes tracing redactability configurable and adds a cluster
setting to enable fine-grained redactability that we can use for
multi-tenant deployments.

The desired outcome is to preserve redactability for tenant traces while
preserving a non-redactable code path for requests using the system
tenant. This preserves existing functionality in production while
keeping the current performance hit of redaction from propagating to
non-tenant deployments.

This PR brings `master` in parity with changes in `release-21.2` via #73117.

Benchmarking this against master results in the following improvements.
old.txt is using `master` @ 65920cf
new.txt is using this commit @ 7c768ac
Benchmark was run using `gceworker`.

We gain back a significant portion of the performance hit:
```
david@davidh:~/go/src/github.com/cockroachdb/cockroach$ benchstat old.txt new.txt
name                                              old time/op    new time/op    delta
Tracing/1node/scan/trace=100%/threshold=1ms-24      1.20ms ± 4%    1.15ms ± 2%   -4.83%  (p=0.000 n=10+9)
Tracing/1node/insert/trace=100%/threshold=1ms-24    1.51ms ± 3%    1.40ms ± 2%   -6.71%  (p=0.000 n=10+10)
Tracing/3node/scan/trace=100%/threshold=1ms-24      2.17ms ± 3%    2.10ms ± 4%   -3.26%  (p=0.003 n=10+10)
Tracing/3node/insert/trace=100%/threshold=1ms-24    2.41ms ± 3%    2.30ms ± 3%   -4.45%  (p=0.000 n=8+9)

name                                              old alloc/op   new alloc/op   delta
Tracing/1node/scan/trace=100%/threshold=1ms-24       115kB ± 1%     120kB ± 1%   +4.39%  (p=0.000 n=10+10)
Tracing/1node/insert/trace=100%/threshold=1ms-24     149kB ± 1%     152kB ± 0%   +1.92%  (p=0.000 n=10+9)
Tracing/3node/scan/trace=100%/threshold=1ms-24       276kB ± 3%     282kB ± 3%   +2.18%  (p=0.035 n=10+10)
Tracing/3node/insert/trace=100%/threshold=1ms-24     265kB ± 1%     264kB ± 2%     ~     (p=0.963 n=8+9)

name                                              old allocs/op  new allocs/op  delta
Tracing/1node/scan/trace=100%/threshold=1ms-24       1.68k ± 0%     1.47k ± 0%  -12.16%  (p=0.000 n=9+9)
Tracing/1node/insert/trace=100%/threshold=1ms-24     2.05k ± 0%     1.76k ± 0%  -14.46%  (p=0.000 n=10+8)
Tracing/3node/scan/trace=100%/threshold=1ms-24       3.10k ± 0%     2.85k ± 0%   -8.21%  (p=0.000 n=10+10)
Tracing/3node/insert/trace=100%/threshold=1ms-24     3.00k ± 0%     2.66k ± 0%  -11.30%  (p=0.000 n=9+10)
```

Resolves cockroachdb#71694

Release note: None

Resolves #71694

Release note: None

tbg committed:
log: add BenchmarkLogEntry_String

Release note: None

tbg committed:
log: add BenchmarkEventf_WithVerboseTrace

Release note: None

tbg committed:
log: optimize logEntry.SafeFormat

via `BenchmarkEventf_WithVerboseTraceSpan`.

```
❯ benchstat before.txt after.txt
name                           old time/op    new time/op    delta
Eventf_WithVerboseTraceSpan-8    5.01µs ± 1%    4.37µs ± 2%  -12.72%  (p=0.000 n=8+9)

name                           old alloc/op   new alloc/op   delta
Eventf_WithVerboseTraceSpan-8    1.20kB ± 0%    1.08kB ± 0%  -10.07%  (p=0.000 n=10+10)

name                           old allocs/op  new allocs/op  delta
Eventf_WithVerboseTraceSpan-8      22.0 ± 0%      16.0 ± 0%  -27.27%  (p=0.000 n=10+10)
```

Release note: None

tbg committed:
tracing: make full redaction of trace messages configurable

When we discovered that tenants could "trace into" KV, we considered
this a security problem since no mechanism ensured that tenants could
receive data they are allowed to see. The KV team reacted to this with
them at the KV-Tenant boundary.

Serverless has shipped with this redactability in place in order to
prevent data leakage across the boundary with KV and now needs to get
back on the `release-21.2` branch to sync with the rest of development.

Unfortunately, making traces redactable had a [significant] performance
overhead. On simple selects, this overhead is in the 10-20% area (when
verbose tracing is enabled).

This presents a stability risk, since some customers run with verbose
tracing enabled for all queries. Once they switch to 21.2, their system
may be underprovisioned as a result of the newly grown tracing overhead
in the `log.{Ve,E}vent{,f}` family of methods. The peformance hit comes
via the additional costs of redacting log strings via the
`redact.Sprintf` function.

This PR changes makes trace redactability a choice. Because we needed to
support mixed-version clusters in #70562 this is not as big a lift as
might have been expected.

The approach taken is that tracers and the spans they generate can be
marked as "redactable". If the flag is enabled, logging events to the
span will use full redaction via `redact.Sprintf`. If the flag is not
enabled a coarse redation of the entire message is done instead. A
further commit will enable redactability on tenant-bound traces since we
do want to maintain that functionality in multi-tenant deployments.

Original issue investigation that kicked off concerns about performance
impact of redaction:
https://github.com/cockroachdb/cockroach/issues/70110#issuecomment-942400586

Touches #70562.

Release note: None

tbg committed:
log: improve Event benchmark to check redactability as well

More CPU and a slightly different allocation profile, though the
allocations are largely driven by the arguments passed to `Eventf` and
they are trivial in this benchmark.

```
go test ./pkg/util/log -run - -bench BenchmarkEvent -benchtime 3s
goos: linux
goarch: amd64
pkg: github.com/cockroachdb/cockroach/pkg/util/log
cpu: AMD Ryzen 7 3800X 8-Core Processor
BenchmarkEventf_WithVerboseTraceSpan/redactable=false-16   759444  4903 ns/op  961 B/op   18 allocs/op
BenchmarkEventf_WithVerboseTraceSpan/redactable=true-16    561772  6831 ns/op  1081 B/op  16 allocs/op
```

Release note: None

dhartunian committed:
server, tracing: add cluster setting to enable tracer redactability

Prior commits added the ability to set a flag on the Tracer to determine
whether Span's should be marked redactable or not. This commit adds
a cluster setting that will enable fine-grained trace redactability. The
intended use is for multi-tenant deployments where KV traces should be
redacted prior to being returned to tenants. Enabling the flag will enable
fine-grained redaction so that tenant operators can continue to debug
their queries via redacted kv trace information instead of having that
information stripped out.

The desired outcome is to preserve redactability for tenant traces while
preserving a non-redactable code path for requests using the system
tenant. This preserves existing functionality in production while
keeping the current performance hit of fine grained redaction from
propagating to non-tenant deployments.

Benchmarking this against master results in the following improvements.
old.txt is using `master` @ 65920cfc71
new.txt is using this commit @ 7c768ac17d
Benchmark was run using `gceworker`.

We gain back a significant portion of the performance hit:
```
david@davidh:~/go/src/github.com/cockroachdb/cockroach$ benchstat old.txt new.txt
name                                              old time/op    new time/op    delta
Tracing/1node/scan/trace=100%/threshold=1ms-24      1.20ms ± 4%    1.15ms ± 2%   -4.83%  (p=0.000 n=10+9)
Tracing/1node/insert/trace=100%/threshold=1ms-24    1.51ms ± 3%    1.40ms ± 2%   -6.71%  (p=0.000 n=10+10)
Tracing/3node/scan/trace=100%/threshold=1ms-24      2.17ms ± 3%    2.10ms ± 4%   -3.26%  (p=0.003 n=10+10)
Tracing/3node/insert/trace=100%/threshold=1ms-24    2.41ms ± 3%    2.30ms ± 3%   -4.45%  (p=0.000 n=8+9)

name                                              old alloc/op   new alloc/op   delta
Tracing/1node/scan/trace=100%/threshold=1ms-24       115kB ± 1%     120kB ± 1%   +4.39%  (p=0.000 n=10+10)
Tracing/1node/insert/trace=100%/threshold=1ms-24     149kB ± 1%     152kB ± 0%   +1.92%  (p=0.000 n=10+9)
Tracing/3node/scan/trace=100%/threshold=1ms-24       276kB ± 3%     282kB ± 3%   +2.18%  (p=0.035 n=10+10)
Tracing/3node/insert/trace=100%/threshold=1ms-24     265kB ± 1%     264kB ± 2%     ~     (p=0.963 n=8+9)

name                                              old allocs/op  new allocs/op  delta
Tracing/1node/scan/trace=100%/threshold=1ms-24       1.68k ± 0%     1.47k ± 0%  -12.16%  (p=0.000 n=9+9)
Tracing/1node/insert/trace=100%/threshold=1ms-24     2.05k ± 0%     1.76k ± 0%  -14.46%  (p=0.000 n=10+8)
Tracing/3node/scan/trace=100%/threshold=1ms-24       3.10k ± 0%     2.85k ± 0%   -8.21%  (p=0.000 n=10+10)
Tracing/3node/insert/trace=100%/threshold=1ms-24     3.00k ± 0%     2.66k ± 0%  -11.30%  (p=0.000 n=9+10)
```

Resolves #71694

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73405)
<!-- Reviewable:end -->

dhartunian commented:
@andreimatei this addresses most of your concerns from #73117 and is a much smaller PR. I will update the one on `release-21.2` tomorrow and respond to the open threads.

You were correct that the use of `redact.Sprintf` vs `fmt.Fprintf` in `spanInner.Recordf` did not significantly affect the runtime. I ran a narrower benchmark (`BenchmarkEventf_WithVerboseTraceSpan`) and found that the largest performance gain by far is in using the `LegacyString` method on the `logEntry`, which is now renamed to `FastString`.

dhartunian commented:
@andreimatei also updated PR msg w/ benchmarks

abarganier reviewed (commented):
Looks good to me, but will defer to Andrei for any stamps. I did have a quick question though re: redactability for system tenants

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-:-MqQ-kFZ3BW1-xQURq8T:b-kwiy0b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @dhartunian)

___
*[pkg/server/node.go, line 1059 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqPzo3yDivdnGaRqGDW:-MqPzo3yDivdnGaRqGDX:b4yw38r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddcb9b0f0fcb4681c056ac655dc246bf75020230/pkg/server/node.go#L1059)):*
> ```
> tr.SetRedactable(true)
> ```
Do we want redactability to be turned off when the request is made for a system tenant? Does this not configure the "node-global" tracer to `redactable = true` without any way for it to be turned back off for system tenants? 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (changes requested):
> Reviewed 1 of 22 files at r7, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73405)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @andreimatei, and @dhartunian)

___
*[-- commits, line 71 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQHcqz6V8GZuf8uXkO):*
something here
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQM2bZ-JXouwK2tW1u:-MqQM2b_Ey9rL61T-l5K:b1vtr57) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L40)):*
> ```Go
> }
> 
> func testTenantTracesAreRedactedImpl(t *testing.T, redactable bool) {
> ```

please comment on the `redactable` arg, explaining that when it's not set, you don't get any messages in the trace
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 49 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQIdrNB09EvBTrcNyP:-MqQIdrNB09EvBTrcNyQ:b7o2pf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L49)):*
> ```Go
> 	)
> 
> 	getTrace := func(t *testing.T, db *gosql.DB) [][]string {
> ```

I think this test would look a lot better if it used the `WithStatementTrace` testing knob instead of session tracing. Feel free to address in a different PR if you prefer.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 75 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQJCIO496pZS6Myc6r:-MqQJCIO496pZS6Myc6s:bk744lc) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L75)):*
> ```Go
> 	defer tc.Stopper().Stop(ctx)
> 
> 	t.Run("system-tenant", func(t *testing.T) {
> ```

please add a comment stating that queries from the system tenant get unredacted traces, and perhaps hint at the mechanism
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 100 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQJu1j-BuYrJgxhC_r:-MqQJu1j-BuYrJgxhC_s:br0hm90) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L100)):*
> ```Go
> 		results := getTrace(t, tenDB)
> 
> 		const redactedMarkerString = "verbose trace message redacted"
> ```

please elevate this to a constant somewhere shared with the production code. Take the opportunity to rename it away from  `sRedactedMarker`. I don't know who started with these type prefixes.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 100 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQKl_X1A1kxURW0Uk0:-MqQKl_X1A1kxURW0Uk1:b-2u5wgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L100)):*
> ```Go
> 		results := getTrace(t, tenDB)
> 
> 		const redactedMarkerString = "verbose trace message redacted"
> ```

it's not clear to me when this "message redacted" string is expected exactly. `redactRecordingForTenant` only seems to use it when processing `DeprecatedFields`, which is supposed to go away. And even then, only when `sp.RedactableLogs` is set.
I'm not sure how the test is passing; how come `DeprecatedFields` arebeing used to render the trace, instead of the `Message` field?
___
*[pkg/util/log/clog_test.go, line 780 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQH2egCcQ8lKUmiZvR:-MqQH2egCcQ8lKUmiZvS:bfoopei) ([raw file](https://github.com/cockroachdb/cockroach/blob/acd4158de1a8be392414aff6565e9a7d591712bc/pkg/util/log/clog_test.go#L780)):*
> ```Go
> 	b.ResetTimer()
> 	defer sp.Finish()
> 	sp.SetVerbose(true)
> ```

nit: move this above `ResetTimer`
___
*[pkg/util/log/log_entry.go, line 133 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQMUWh82WJLvKY8UaU:-MqQMUWh82WJLvKY8UaV:blqnju1) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/log_entry.go#L133)):*
> ```Go
> }
> 
> // FastString is an old implementation of String(). It is still around
> ```

well then why don't we just make this be `logEntry.String()` if it's faster? If you do that, put a comment on it than it doesn't follow the usual `redact.StringWithoutMarkers(e)` pattern for `String()` implementations of `SafeFromatters` because performance.

But also see my comments on `eventInternal`. I think if you want to do all I describe there, this method does not need to exist at all (although it might be a good idea to make `logEntry.String()` fast regardless of the current PR).
___
*[pkg/util/log/log_entry.go, line 147 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQN1kx9mmnrEuw-SiA:-MqQN1kx9mmnrEuw-SiB:b-xs9614) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/log_entry.go#L147)):*
> ```Go
> 		buf.WriteString(entry.File)
> 		buf.WriteByte(':')
> 		// TODO(knz): The "canonical" way to represent a file/line prefix
> ```

if this method stays different from `logEntry.String`, then considerations about SQL don't apply, do they? And I don't understand what this comment is referring to anyway. But on the other hand, I do not want a trailing colon anyway. I'd get rid of this comment.
___
*[pkg/util/log/trace.go, line 104 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQSfKzBsNBYKZTTTyl:-MqQSfKzBsNBYKZTTTym:b-bhnwec) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L104)):*
> ```
> may or may not be redactable
> ```

consider linking to logEntry.redactable.
___
*[pkg/util/log/trace.go, line 105 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQQl5i3BD-XJWo2IoQ:-MqQQl5i3BD-XJWo2IoR:bwsa164) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L105)):*
> ```Go
> // eventInternal is the common code for logging an event to trace and/or event
> // logs. Entries passed to this method may or may not be redactable. However,
> // if sp.Redactable() is true then ideally they are for best results.
> ```

these "best results" need more words, I think. Say that a non-redactable logEntry will be completely redacted if `sp.Redactable()`
___
*[pkg/util/log/trace.go, line 106 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQPupR77wsn3ZVFRSq:-MqQPupR77wsn3ZVFRSr:b-h59c3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L106)):*
> ```Go
> // logs. Entries passed to this method may or may not be redactable. However,
> // if sp.Redactable() is true then ideally they are for best results.
> func eventInternal(sp *tracing.Span, el *ctxEventLog, isErr bool, entry *logEntry) {
> ```

Should `spanInner.Recordf()` use `redact.Sprintf()` or `fmt.Sprintf()` based on `sp.Redactable()`? I think that's a good idea by itself, but in addition, if you do that and also rename `logEntry.FastString()` as `String(), this code wouldn't need this branch. This code in the logging package would then not need to care about `sp.Redactable()` twice - once when preparing the entry, once here.

(this second part feel free to leave for later if you don't want to bite)
in fact, I think the code in the `logging` package should not care about `sp.Redactable()` at all. I don't think `sp.Redactable()` should be exported (or, in fact, exist at all as a method). I think we can get there by having the `log.Eventf()` code and its friends pass the format and the args to `sp.Recordf()`, not a `logEntry`.
This code smells:
```
		entry := makeUnstructuredEntry(ctx,
			severity.INFO, /* unused for trace events */
			channel.DEV,   /* unused for trace events */
			depth+1,
			sp.Redactable(),
			format, args...)
		eventInternal(sp, el, isErr, &entry)
```

Those "unused for trace events" comments hint that something's wrong with the layering. I think `eventInternal` should take the format, args, and depth, do whatever for the `eventLog` case (I don't care), and pass them all to `Recordf()` (perhaps it can terminate the `depth` and prepending the file:line to the format).
___
*[pkg/util/log/trace.go, line 115 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQO0tT8KV-KY6re1BY:-MqQO0tT8KV-KY6re1BZ:b-epk7l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L115)):*
> ```Go
> 		} else {
> 			// FastString is significantly faster than String since the latter
> 			// round-trips through redaction. See BenchmarkEventf_WithVerboseTraceSpan.
> ```

this comment belongs around `FastString`
___
*[pkg/util/tracing/tracer.go, line 175 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQOzpx7MbR1Ei6Kasg:-MqQOzpx7MbR1Ei6Kash:b-kop5n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/tracing/tracer.go#L175)):*
> ```Go
> 	// as redactable. This will make unstructured events logged to those
> 	// spans redactable.
> 	// Currently, this is used to mark tenant-bound spans as redactable
> ```

this comment suggests that "tenant-bound spans" are somehow different from other spans, but that's not true, is it?
___
*[pkg/util/tracing/tracer.go, line 314 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQP9le6HWDnWunA6Ne:-MqQP9le6HWDnWunA6Nf:b-cvzx4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/tracing/tracer.go#L314)):*
> ```Go
> }
> 
> // Redactable returns true if the tracer is configured to emit
> ```

it also affects spans that have already been opened, hasn't it? If anything, this should talk about future collection of recordings.
___
*[pkg/util/tracing/tracingpb/recorded_span.proto, line 75 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQLKs2-ZABc9oQaj1W:-MqQLKs2-ZABc9oQaj1X:b-sy18w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/tracing/tracingpb/recorded_span.proto#L75)):*
> ```protobuf
> 
>   // RedactableLogs determines whether the verbose log messages are redactable.
>   // This field was introduced in the 22.1 cycle. It can be removed in the 22.2
> ```

please remove the word "cycle" from here; it's very confusing.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73405)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @andreimatei, and @dhartunian)

___
*[pkg/util/log/trace.go, line 106 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQPupR77wsn3ZVFRSq:-MqQXoCg1xK5GNr5UQI-:bgngmgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L106)):*
> (perhaps it can terminate the depth and prepending the file:line to the format)

If we do all this, and also push the handling of the `depth` inside `sp.Recordf`, then the logging code could also stop caring about `sp.IsVerbose()` in `getSpanOrEventLog()`, which would be another win in my book. The logging code could blindly delegate all the work to the span, and a non-verbose span could short-circuit cheaply.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73405)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @andreimatei, and @dhartunian)

___
*[-- commits, line 71 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQHcqz6V8GZuf8uXkO):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

something here
</blockquote></details>

done.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQM2bZ-JXouwK2tW1u:-MqQcgPPE0OmtgmPRt8K:b-dcdlb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L40)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please comment on the `redactable` arg, explaining that when it's not set, you don't get any messages in the trace
</blockquote></details>

This arg is no longer necessary because we don't test 
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 49 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQIdrNB09EvBTrcNyP:-MqQha458X8t8VpCpn6Z:b53cdma) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L49)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this test would look a lot better if it used the `WithStatementTrace` testing knob instead of session tracing. Feel free to address in a different PR if you prefer.
</blockquote></details>

Will do separately.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 75 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQJCIO496pZS6Myc6r:-MqQhlf37K3D2sqa3pEU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L75)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment stating that queries from the system tenant get unredacted traces, and perhaps hint at the mechanism
</blockquote></details>

Done.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 100 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQJu1j-BuYrJgxhC_r:-MqQXuqp0gbvT7FMxkRr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L100)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please elevate this to a constant somewhere shared with the production code. Take the opportunity to rename it away from  `sRedactedMarker`. I don't know who started with these type prefixes.
</blockquote></details>

Done.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 100 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQKl_X1A1kxURW0Uk0:-MqQlneVBnAKyJNyDn2q:bnuq22j) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L100)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it's not clear to me when this "message redacted" string is expected exactly. `redactRecordingForTenant` only seems to use it when processing `DeprecatedFields`, which is supposed to go away. And even then, only when `sp.RedactableLogs` is set.
I'm not sure how the test is passing; how come `DeprecatedFields` arebeing used to render the trace, instead of the `Message` field?
</blockquote></details>

This test was originally written to deal with the case of dropping tenant traces. Now that we redact and keep them, the expectation is that we *don't* see that redacted marker in the results.
___
*[pkg/server/node.go, line 1059 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqPzo3yDivdnGaRqGDW:-MqQmItXCbmdmUYI03Nn:bft2va7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddcb9b0f0fcb4681c056ac655dc246bf75020230/pkg/server/node.go#L1059)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

> ```
> tr.SetRedactable(true)
> ```
Do we want redactability to be turned off when the request is made for a system tenant? Does this not configure the "node-global" tracer to `redactable = true` without any way for it to be turned back off for system tenants? 
</blockquote></details>

True. Since Serverless is the only deployment that will pay this cost, and is paying it system-wide already we're kepping this as a one-way switch unless you restart the node. I am happy to add further configuration knows as follow-ups.
___
*[pkg/util/log/clog_test.go, line 780 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQH2egCcQ8lKUmiZvR:-MqQY3yd6cw6oCtk148T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/acd4158de1a8be392414aff6565e9a7d591712bc/pkg/util/log/clog_test.go#L780)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: move this above `ResetTimer`
</blockquote></details>

Done.
___
*[pkg/util/log/log_entry.go, line 133 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQMUWh82WJLvKY8UaU:-MqQmUBO1Q6bRNdlY5Qo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/log_entry.go#L133)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well then why don't we just make this be `logEntry.String()` if it's faster? If you do that, put a comment on it than it doesn't follow the usual `redact.StringWithoutMarkers(e)` pattern for `String()` implementations of `SafeFromatters` because performance.

But also see my comments on `eventInternal`. I think if you want to do all I describe there, this method does not need to exist at all (although it might be a good idea to make `logEntry.String()` fast regardless of the current PR).
</blockquote></details>

Done.
___
*[pkg/util/log/log_entry.go, line 147 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQN1kx9mmnrEuw-SiA:-MqQmchMA-OsuEKsBEYT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/log_entry.go#L147)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if this method stays different from `logEntry.String`, then considerations about SQL don't apply, do they? And I don't understand what this comment is referring to anyway. But on the other hand, I do not want a trailing colon anyway. I'd get rid of this comment.
</blockquote></details>

Done.
___
*[pkg/util/log/trace.go, line 104 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQSfKzBsNBYKZTTTyl:-MqQbNH54AoeM1C3zCcd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L104)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> ```
> may or may not be redactable
> ```

consider linking to logEntry.redactable.

</blockquote></details>

Done.
___
*[pkg/util/log/trace.go, line 105 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQQl5i3BD-XJWo2IoQ:-MqQbNYH6eOL62vlgkFM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L105)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

these "best results" need more words, I think. Say that a non-redactable logEntry will be completely redacted if `sp.Redactable()`
</blockquote></details>

Done.
___
*[pkg/util/log/trace.go, line 106 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQPupR77wsn3ZVFRSq:-MqQaJ8t1njGnfCD7Sx6:b-1vunra) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L106)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> (perhaps it can terminate the depth and prepending the file:line to the format)

If we do all this, and also push the handling of the `depth` inside `sp.Recordf`, then the logging code could also stop caring about `sp.IsVerbose()` in `getSpanOrEventLog()`, which would be another win in my book. The logging code could blindly delegate all the work to the span, and a non-verbose span could short-circuit cheaply.
</blockquote></details>

Gotcha. Made the first change. I like that.
___
*[pkg/util/log/trace.go, line 115 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQO0tT8KV-KY6re1BY:-MqQaFwh7YMELxE2nEX2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/log/trace.go#L115)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment belongs around `FastString`
</blockquote></details>

Done.
___
*[pkg/util/tracing/tracer.go, line 175 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQOzpx7MbR1Ei6Kasg:-MqQbYz2CW23K9PjCrt_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/tracing/tracer.go#L175)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment suggests that "tenant-bound spans" are somehow different from other spans, but that's not true, is it?
</blockquote></details>

Done.
___
*[pkg/util/tracing/tracer.go, line 314 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQP9le6HWDnWunA6Ne:-MqQcQawAn2hOxDCn16X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/tracing/tracer.go#L314)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it also affects spans that have already been opened, hasn't it? If anything, this should talk about future collection of recordings.
</blockquote></details>

Done.
___
*[pkg/util/tracing/tracingpb/recorded_span.proto, line 75 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQLKs2-ZABc9oQaj1W:-MqQcW9H19CSriqPtHws:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/util/tracing/tracingpb/recorded_span.proto#L75)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please remove the word "cycle" from here; it's very confusing.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

but I kinda insist on making `tr.redactable` be based on more static configuration, rather than setting it on RPCs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-:-MqQzBUPA2B2VI50GoQB:b-cpz7fw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @andreimatei, and @dhartunian)

___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQM2bZ-JXouwK2tW1u:-MqQuvwrE38DPejz0E5J:b-af2t6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L40)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This arg is no longer necessary because we don't test 
</blockquote></details>

did you backtrack this change in the wrong commit?
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 116 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQvFSD4xDACnwqVLY3:-MqQvFSE9Y8VcncBBGmk:btyap15) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L116)):*
> ```Go
> 		require.True(t, found, "did not see expected trace message '%q':\n%s",
> 			visibleString, sqlutils.MatrixToStr(results))
> 		require.False(t, foundRedactedMarker, "unexpectedly found '%q':\n%s",
> ```

I'm still confused about this check. What is its point? That marker has something to do with old versions, right?
___
*[pkg/server/node.go, line 1058 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQwcN95QxmVdoYflt1:-MqQwcN95QxmVdoYflt2:br0n8ac) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/server/node.go#L1058)):*
> ```Go
> 	// resolved via #58610, this code can be removed so that all traces
> 	// have redactability enabled.
> 	if !tr.Redactable() && tenID != roachpb.SystemTenantID {
> ```

I imagined that redactability would be more static configuration (flag, or cluster setting).
I figure there must be something else telling a kv node that it's part of a multi-tenant cluster, and we'd piggy back on that.
___
*[pkg/server/node_tenant_test.go, line 51 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQxHVX5xAHrD1mjUMg:-MqQxHVX5xAHrD1mjUMh:bn7vnni) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/server/node_tenant_test.go#L51)):*
> ```Go
> 		tracer.SetRedactable(true)
> 		ctx, sp := tracer.StartSpanCtx(ctx, "foo", tracing.WithRecording(tracing.RecordingVerbose))
> 		sp.SetVerbose(true)
> ```

we shouldn't need `SetVerbose()`, we started recording already above when we created the span.
___
*[pkg/util/tracing/span_inner.go, line 189 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQy7Qc9xSL_npwGTNj:-MqQy7Qc9xSL_npwGTNk:b-tmf33z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/util/tracing/span_inner.go#L189)):*
> ```Go
> 		str = redact.Sprintf(format, args...)
> 	} else {
> 		// Format is considered safe but here we want to make sure the
> ```

`s/Format/format` (to refer to the variable)?
I'm not sure what the first part of this comment is telling me, and why there's a "but". I suspect other readers will be confused too.

This comment should spell out that `fmt.Sprintf` is faster than `redact.Sprintf()`, and it should particularly mention `logEntry`.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73405)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @andreimatei, and @dhartunian)

___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQM2bZ-JXouwK2tW1u:-MqRaGmL0xUn6KlENRqD:b-yw9wix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71ec9ce7cdebbb0a2adfc384091233da1ce83f0a/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L40)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

did you backtrack this change in the wrong commit?
</blockquote></details>

yeah I messed something up. Going to restore the full test in the PR since tenant-level redaction is now configurable as well.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go, line 116 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQvFSD4xDACnwqVLY3:-MqR2Tz9-u_g72vbz7eh:b-lp4dey) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/ccl/kvccl/kvtenantccl/tenant_trace_test.go#L116)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm still confused about this check. What is its point? That marker has something to do with old versions, right?
</blockquote></details>

I added an explanation of why we would never expect this to show up. Seems like it will be appropriate to remove once we remove the `TraceRedactedMarker` altogether.
___
*[pkg/server/node.go, line 1059 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqPzo3yDivdnGaRqGDW:-MqR3-UQ2L3aRQfCk4HD:b-ub870b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddcb9b0f0fcb4681c056ac655dc246bf75020230/pkg/server/node.go#L1059)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

True. Since Serverless is the only deployment that will pay this cost, and is paying it system-wide already we're kepping this as a one-way switch unless you restart the node. I am happy to add further configuration knows as follow-ups.
</blockquote></details>

Added a boolean cluster setting for this which should alleviate concerns.
___
*[pkg/server/node.go, line 1058 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQwcN95QxmVdoYflt1:-MqR2yFb--CfB5nFZnRP:bjrf6ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/server/node.go#L1058)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I imagined that redactability would be more static configuration (flag, or cluster setting).
I figure there must be something else telling a kv node that it's part of a multi-tenant cluster, and we'd piggy back on that.
</blockquote></details>

Added boolean cluster setting.
___
*[pkg/server/node_tenant_test.go, line 51 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQxHVX5xAHrD1mjUMg:-MqRC90ZD_ycW3zA7oHr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/server/node_tenant_test.go#L51)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

we shouldn't need `SetVerbose()`, we started recording already above when we created the span.
</blockquote></details>

Done.
___
*[pkg/util/tracing/span_inner.go, line 189 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-MqQy7Qc9xSL_npwGTNj:-MqRBvU5A9CnlAJv6bnm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c37e645b3b1ebade6e52932d646115a49bd20e3/pkg/util/tracing/span_inner.go#L189)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`s/Format/format` (to refer to the variable)?
I'm not sure what the first part of this comment is telling me, and why there's a "but". I suspect other readers will be confused too.

This comment should spell out that `fmt.Sprintf` is faster than `redact.Sprintf()`, and it should particularly mention `logEntry`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dhartunian commented:
bors r=andreimatei

knz reviewed (approved):
Third commit: :kiss: 
(but also :facepalm: for the first time around when the code was written)

> Reviewed 14 of 14 files at r17, 1 of 1 files at r18, 1 of 1 files at r19, 10 of 10 files at r20, 1 of 1 files at r21, 5 of 6 files at r22, 1 of 1 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73405#-:-MqUnfaR69UiIknZpaZg:b9dnn8z)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @andreimatei, @dhartunian, and @knz)



<!-- Sent from Reviewable.io -->
